File: sdks/typescript/tests/test_evaris.ts
/**
 * Tests for EvarisExporter and init()/score() one-liners.
 *
 * Matches Python test_evaris.py (15 tests).
 */
import { describe, it, expect, vi, afterEach } from "vitest";
import { EvarisExporter, DEFAULT_ENDPOINT } from "../src/evaris/exporter.js";
import { shutdown, trace } from "../src/index.js";
import { SpanExportResult } from "../src/exporter.js";
import type { Span } from "../src/types.js";
import * as evarisClient from "../src/evaris/client.js";

afterEach(() => {
  vi.restoreAllMocks();
  shutdown();
});

function makeSpan(overrides: Partial<Span> = {}): Span {
  return {
    traceId: "a".repeat(32),
    spanId: "b".repeat(16),
    name: "test_span",
    kind: "tool",
    status: "ok",
    startedAt: new Date().toISOString(),
    endedAt: new Date().toISOString(),
    durationNs: 1_000_000,
    ...overrides,
  };
}

// ── EvarisExporter ───────────────────────────────────────────────────────────

describe("EvarisExporter", () => {
  it("POSTs spans to Evaris backend", async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      statusText: "OK",
    });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "test-key", projectId: "proj-1" });
    const result = await exporter.export([makeSpan()]);

    expect(result).toBe(SpanExportResult.SUCCESS);
    expect(mockFetch).toHaveBeenCalledOnce();
    const url = mockFetch.mock.calls[0][0] as string;
    expect(url).toContain("runtime.evaris.ai");
  });

  it("includes authorization header", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "sk-test-123", projectId: "proj-1" });
    await exporter.export([makeSpan()]);

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const headers = opts.headers as Record<string, string>;
    expect(headers["Authorization"]).toBe("Bearer sk-test-123");
  });

  it("includes project ID header", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-42" });
    await exporter.export([makeSpan()]);

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const headers = opts.headers as Record<string, string>;
    expect(headers["X-Evaris-Project-ID"]).toBe("proj-42");
  });

  it("serializes span as JSON in request body", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    const span = makeSpan({ name: "my_span", kind: "llm" });
    await exporter.export([span]);

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const body = JSON.parse(opts.body as string);
    expect(Array.isArray(body)).toBe(true);
    expect(body[0].name).toBe("my_span");
    expect(body[0].kind).toBe("llm");
  });

  it("returns FAILURE on network error", async () => {
    const mockFetch = vi.fn().mockRejectedValue(new Error("network error"));
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    const result = await exporter.export([makeSpan()]);
    expect(result).toBe(SpanExportResult.FAILED);
  });

  it("returns FAILURE on HTTP error", async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: false,
      status: 500,
      statusText: "Internal Server Error",
    });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    const result = await exporter.export([makeSpan()]);
    expect(result).toBe(SpanExportResult.FAILED);
  });

  it("uses custom endpoint", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({
      apiKey: "key",
      projectId: "proj-1",
      endpoint: "https://custom.example.com/v1/traces",
    });
    await exporter.export([makeSpan()]);

    const url = mockFetch.mock.calls[0][0] as string;
    expect(url).toBe("https://custom.example.com/v1/traces");
  });

  it("serializes usage in span body", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    const span = makeSpan();
    span.usage = { input: 100, output: 50, total: 150, totalCost: 0.005 };
    await exporter.export([span]);

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const body = JSON.parse(opts.body as string);
    expect(body[0].usage.input).toBe(100);
    expect(body[0].usage.total_cost).toBe(0.005);
  });

  it("serializes error in span body", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    const span = makeSpan({ status: "error" });
    span.error = { message: "boom", type: "ValueError" };
    await exporter.export([span]);

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const body = JSON.parse(opts.body as string);
    expect(body[0].error.message).toBe("boom");
    expect(body[0].error.type).toBe("ValueError");
  });

  it("shutdown is a safe no-op", () => {
    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    expect(() => exporter.shutdown()).not.toThrow();
  });

  it("forceFlush returns true", () => {
    const exporter = new EvarisExporter({ apiKey: "key", projectId: "proj-1" });
    expect(exporter.forceFlush()).toBe(true);
  });
});

// ── DEFAULT_ENDPOINT ─────────────────────────────────────────────────────────

describe("DEFAULT_ENDPOINT", () => {
  it("points to runtime.evaris.ai", () => {
    expect(DEFAULT_ENDPOINT).toBe("https://runtime.evaris.ai/v1/traces");
  });
});

// ── evaris.init() ────────────────────────────────────────────────────────────

describe("evaris.init()", () => {
  it("configures etrace with EvarisExporter", async () => {
    const mockFetch = vi.fn().mockResolvedValue({ ok: true, status: 200, statusText: "OK" });
    vi.stubGlobal("fetch", mockFetch);

    evarisClient.init({
      apiKey: "test-key",
      projectId: "proj-1",
      autoInstrument: { llm: false },
    });

    trace("test", () => {});

    // Allow async export to complete
    await new Promise((r) => setTimeout(r, 10));
    shutdown();

    expect(mockFetch).toHaveBeenCalled();
  });
});

// ── evaris.score() ───────────────────────────────────────────────────────────

describe("evaris.score()", () => {
  it("POSTs score to Evaris backend", async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: () => Promise.resolve({ id: "score-1" }),
    });
    vi.stubGlobal("fetch", mockFetch);

    await evarisClient.score({
      apiKey: "key",
      projectId: "proj-1",
      traceId: "abc123",
      name: "relevance",
      value: 0.9,
    });

    expect(mockFetch).toHaveBeenCalledOnce();
    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const body = JSON.parse(opts.body as string);
    expect(body.trace_id).toBe("abc123");
    expect(body.name).toBe("relevance");
    expect(body.value).toBe(0.9);
  });

  it("includes auth headers in score request", async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: () => Promise.resolve({ id: "score-1" }),
    });
    vi.stubGlobal("fetch", mockFetch);

    await evarisClient.score({
      apiKey: "sk-test",
      projectId: "proj-1",
      traceId: "trace-abc",
      name: "accuracy",
      value: 1.0,
    });

    const opts = mockFetch.mock.calls[0][1] as RequestInit;
    const headers = opts.headers as Record<string, string>;
    expect(headers["Authorization"]).toBe("Bearer sk-test");
    expect(headers["X-Evaris-Project-ID"]).toBe("proj-1");
  });
});

File: sdks/python/tests/test_evaris.py
"""Tests for etrace-evaris plugin.

Tests the EvarisExporter and the init() one-liner that configures
etrace with Evaris cloud backend credentials.
"""

from __future__ import annotations

import json
from unittest.mock import MagicMock, patch

import etrace
from etrace._exporter import SpanExportResult
from etrace._types import TraceKind, TraceStatus


class TestEvarisExporter:
    """Unit tests for EvarisExporter."""

    def _make_span(self, **overrides):
        from datetime import UTC, datetime

        from etrace._types import Span

        defaults = dict(
            trace_id="a" * 32,
            span_id="b" * 16,
            name="test_span",
            kind=TraceKind.TOOL,
            status=TraceStatus.OK,
            started_at=datetime.now(UTC).isoformat(),
            ended_at=datetime.now(UTC).isoformat(),
            duration_ns=1_000_000,
        )
        defaults.update(overrides)
        return Span(**defaults)

    def test_export_sends_to_evaris_backend(self):
        """EvarisExporter POSTs spans to runtime.evaris.ai."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.raise_for_status = MagicMock()

        with patch("httpx.post", return_value=mock_response) as mock_post:
            exporter = EvarisExporter(api_key="test-key", project_id="proj-1")
            span = self._make_span()
            result = exporter.export([span])

            assert result == SpanExportResult.SUCCESS
            mock_post.assert_called_once()
            call_kwargs = mock_post.call_args
            assert "runtime.evaris.ai" in call_kwargs.kwargs.get("url", "") or "runtime.evaris.ai" in str(call_kwargs)

    def test_export_includes_auth_headers(self):
        """EvarisExporter includes api_key in request headers."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.raise_for_status = MagicMock()

        with patch("httpx.post", return_value=mock_response) as mock_post:
            exporter = EvarisExporter(api_key="sk-test-123", project_id="proj-1")
            span = self._make_span()
            exporter.export([span])

            headers = mock_post.call_args.kwargs.get("headers", {})
            assert (
                "Authorization" in headers
                or "X-API-Key" in headers
                or any("sk-test-123" in str(v) for v in headers.values())
            )

    def test_export_includes_project_id_header(self):
        """EvarisExporter includes project_id in request headers."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.raise_for_status = MagicMock()

        with patch("httpx.post", return_value=mock_response) as mock_post:
            exporter = EvarisExporter(api_key="key", project_id="proj-42")
            span = self._make_span()
            exporter.export([span])

            headers = mock_post.call_args.kwargs.get("headers", {})
            assert "proj-42" in str(headers)

    def test_export_sends_span_as_json(self):
        """EvarisExporter serializes spans to JSON in request body."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.raise_for_status = MagicMock()

        with patch("httpx.post", return_value=mock_response) as mock_post:
            exporter = EvarisExporter(api_key="key", project_id="proj-1")
            span = self._make_span(name="my_span", kind=TraceKind.LLM)
            exporter.export([span])

            content = mock_post.call_args.kwargs.get("content", "")
            if not content:
                json_arg = mock_post.call_args.kwargs.get("json", None)
                assert json_arg is not None
            else:
                data = json.loads(content)
                assert isinstance(data, (dict, list))

    def test_export_failure_returns_failure(self):
        """EvarisExporter returns FAILURE when POST fails."""
        from etrace_evaris import EvarisExporter

        with patch("httpx.post", side_effect=Exception("network error")):
            exporter = EvarisExporter(api_key="key", project_id="proj-1")
            span = self._make_span()
            result = exporter.export([span])
            assert result == SpanExportResult.FAILURE

    def test_export_http_error_returns_failure(self):
        """EvarisExporter returns FAILURE on non-200 response."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 500
        mock_response.raise_for_status = MagicMock(side_effect=Exception("500"))

        with patch("httpx.post", return_value=mock_response):
            exporter = EvarisExporter(api_key="key", project_id="proj-1")
            span = self._make_span()
            result = exporter.export([span])
            assert result == SpanExportResult.FAILURE

    def test_custom_endpoint(self):
        """EvarisExporter can use a custom endpoint."""
        from etrace_evaris import EvarisExporter

        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.raise_for_status = MagicMock()

        with patch("httpx.post", return_value=mock_response) as mock_post:
            exporter = EvarisExporter(
                api_key="key",
                project_id="proj-1",
                endpoint="https://custom.example.com/v1/traces",
            )
            span = self._make_span()
            exporter.export([span])

            url = (
                mock_post.call_args.kwargs.get("url", "") or mock_post.call_args.args[0]
                if mock_post.call_args.args
                else ""
            )
            assert "custom.example.com" in url

    def test_shutdown_is_noop(self):
        """EvarisExporter.shutdown() is a safe no-op."""
        from etrace_evaris import EvarisExporter

        exporter = EvarisExporter(api_key="key", project_id="proj-1")
        exporter.shutdown()  # Should not raise

    def test_force_flush_returns_true(self):
        """EvarisExporter.force_flush() returns True (synchronous export)."""
        from etrace_evaris import EvarisExporter

        exporter = EvarisExporter(api_key="key", project_id="proj-1")
        assert exporter.force_flush() is True


class TestEvarisInit:
    """Tests for etrace_evaris.init() one-liner."""

    def test_init_configures_etrace_with_evaris_exporter(self):
        """etrace_evaris.init() calls etrace.init() with EvarisExporter."""
        from etrace_evaris import EvarisExporter

        with patch.object(etrace, "init") as mock_init:
            import etrace_evaris

            etrace_evaris.init(api_key="test-key", project_id="proj-1")

            mock_init.assert_called_once()
            call_kwargs = mock_init.call_args.kwargs
            exporters = call_kwargs.get("exporters", [])
            assert len(exporters) == 1
            assert isinstance(exporters[0], EvarisExporter)

    def test_init_passes_service_name(self):
        """etrace_evaris.init() forwards service_name."""
        with patch.object(etrace, "init") as mock_init:
            import etrace_evaris

            etrace_evaris.init(
                api_key="key",
                project_id="proj-1",
                service_name="my-service",
            )

            call_kwargs = mock_init.call_args.kwargs
            assert call_kwargs["service_name"] == "my-service"

    def test_init_passes_environment(self):
        """etrace_evaris.init() forwards environment."""
        with patch.object(etrace, "init") as mock_init:
            import etrace_evaris

            etrace_evaris.init(
                api_key="key",
                project_id="proj-1",
                environment="staging",
            )

            call_kwargs = mock_init.call_args.kwargs
            assert call_kwargs["environment"] == "staging"

    def test_init_forwards_auto_instrument(self):
        """etrace_evaris.init() forwards auto_instrument setting."""
        with patch.object(etrace, "init") as mock_init:
            import etrace_evaris

            etrace_evaris.init(
                api_key="key",
                project_id="proj-1",
                auto_instrument={"llm": False},
            )

            call_kwargs = mock_init.call_args.kwargs
            assert call_kwargs["auto_instrument"] == {"llm": False}


class TestEvarisScore:
    """Tests for etrace_evaris.score() API."""

    def test_score_posts_to_evaris(self):
        """etrace_evaris.score() POSTs a score to the Evaris backend."""
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = {"id": "score-1"}

        with patch("httpx.post", return_value=mock_response) as mock_post:
            import etrace_evaris

            etrace_evaris.score(
                api_key="key",
                project_id="proj-1",
                trace_id="abc123",
                name="relevance",
                value=0.9,
            )

            mock_post.assert_called_once()

    def test_score_includes_auth_and_trace(self):
        """score() includes api_key, project_id, and trace_id in request."""
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = {"id": "score-1"}

        with patch("httpx.post", return_value=mock_response) as mock_post:
            import etrace_evaris

            etrace_evaris.score(
                api_key="sk-test",
                project_id="proj-1",
                trace_id="trace-abc",
                name="accuracy",
                value=1.0,
            )

            call = mock_post.call_args
            call.kwargs.get("headers", {})
            json_body = call.kwargs.get("json", {})
            assert json_body.get("trace_id") == "trace-abc"
            assert json_body.get("name") == "accuracy"
            assert json_body.get("value") == 1.0
